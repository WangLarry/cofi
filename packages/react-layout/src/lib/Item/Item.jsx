/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import ItemView from '../ItemView';
import MobileItemView from '../MobileItemView';

export const layoutTypes = {
  TABS: 'tabs',
  SCROLL: 'scroll',
  MOBILE: 'mobile',
};

const Item = ({ title, layout, size, selected, sections = [], mainActions, optionsActions }) => {
  const [_rootSectionsRef] = useState(React.createRef());
  const [_selected, setSelected] = useState({}); // { sectionId, elementId }
  const [_mainActions, setMainActions] = useState(); // to keep action.ref between rendered, and popover will be shown
  const [_optionsActions, setOptionsActions] = useState();

  useEffect(() => setSelected(selected || {}), [selected]);

  useEffect(() => setMainActions(mainActions ? prepareActions(mainActions) : undefined), [mainActions]);
  useEffect(() => setOptionsActions(optionsActions ? prepareActions(optionsActions) : undefined), [optionsActions]);

  const evaluatedMainActions = _mainActions ? evaluateActions(_mainActions) : undefined;
  const evaluatedOptionsActions = _optionsActions ? evaluateActions(_optionsActions) : undefined;
  
  useEffect(() => {
    // scroll to element
    if (_selected && _selected.elementId) {
      const sections = _rootSectionsRef.current;
      sections.scrollTo({
        top: sections.querySelector(`#${_selected.elementId}`).offsetTop - sections.offsetTop - 10,
        left: 0,
        behavior: 'smooth',
      });
    }
  }, [_selected, _rootSectionsRef]);

  // filter section by exclude condition
  let evaluatedSections = evaluateSections(sections);

  if (!evaluatedSections.length) return (null);

  switch (layout) {
  case layoutTypes.MOBILE: {
    return (<MobileItemView title={title} sections={evaluatedSections} _mainActions={evaluatedMainActions}
      optionsActions={evaluatedOptionsActions} size={size} />);
  }
  default: {
    const selectedSectionId = _selected.sectionId || evaluatedSections[0].id;

    const onChangeTab = (event, selectedSectionId) => {
      setSelected({ sectionId: selectedSectionId, elementId: undefined });
  
      // if scroll layout - scroll to section
      if (layout === layoutTypes.SCROLL) {
        const section = evaluatedSections.find(x => x.id === selectedSectionId);
        section.ref.current.scrollIntoView({ behavior: 'smooth' });
      }
    };

    // create the tabs menu items from the sections structure
    const tabs = layout ? {
      items: evaluatedSections.map(section => ({ value: section.id, label: section.title })),
      value: selectedSectionId,
      onChange: onChangeTab,
    } : undefined;
    
    // if tabs layout - show only selected tab, without title for first level section
    if (layout === layoutTypes.TABS) {
      evaluatedSections = [{ ...evaluatedSections.find(section => section.id === selectedSectionId), title: undefined }];
    }

    // prepare footer
    const footer = evaluatedMainActions ? { actions: evaluatedMainActions } : undefined;

    // prepare options
    const options = evaluatedOptionsActions ? { actions: evaluatedOptionsActions } : undefined;

    return (<ItemView title={title} tabs={tabs} sections={evaluatedSections} sectionsRef={_rootSectionsRef}
      footer={footer} options={options} size={size} />);
  }
  }
};

Item.propTypes = {
  title: PropTypes.string,
  layout: PropTypes.oneOf(Object.values(layoutTypes)),
  sections: PropTypes.array,
  selected: PropTypes.object,
  mainActions: PropTypes.array,
  optionsActions: PropTypes.array,
  size: PropTypes.number,
};

export default Item;

function evaluateSections(sections = []) {
  const newSections = sections.map(s => ({ ...s, ref: React.createRef() }));
  return newSections.filter((section) => {
    const renderSections = evaluateSections(section.sections);
    if (renderSections.length) {
      section.sections = renderSections;
    }
    return !(section.exclude && section.exclude()) || section.sections;
  });
}

function evaluateActions(actions) {
  // exclude actions, and set disabled on each render
  return actions
    .filter(action => !(action.exclude && action.exclude()))
    .map(action => Object.assign({}, action, {
      disabled: action.disable && action.disable(),
    }));
}

function prepareActions(actions) {
  // apply actions popover
  return actions.map(action => {
    if (action.popover) {
      const ref = React.createRef();
      return { ...action, elementRef: ref, popover: { ...action.popover, targetRef: ref } };
    }
    return action;
  });
}
