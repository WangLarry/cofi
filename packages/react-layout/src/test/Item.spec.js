/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { shallow } from 'enzyme';
import Item from '../lib/Item';
 
describe('Item', () => {
  let props;
 
  beforeEach(() => {
    props = {
      title: 'Employee',
      sections: [{
        id: 'general',
        title: 'General',
      }],
      layout: 'scroll',
      mainActions: [],
      optionsActions: [],
    };
  });
  it('Should render - ok', () => {
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render with layout tabs - ok', () => {
    props.layout = 'tabs';
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render with layout mobile - ok', () => {
    props.layout = 'mobile';
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render without layout - ok', () => {
    props.layout = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render without title - ok', () => {
    props.title = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render with all filtered sections - ok', () => {
    props.sections[0].exclude = () => true;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render without mainActions - ok', () => {
    props.mainActions = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render with mainActions - popover - ok', () => {
    props.mainActions = [{
      label: 'Save & Close',
      type: 'secondary',
      disable: () => false,
      onClick: jest.fn(),
    }, {
      label: 'Save',
      type: 'primary',
      icon: () => null,
      disable: () => false,
      onClick: jest.fn(),
      popover: {
        title: 'Handle Fields',
        open: () => false,
        component: () => (<div>hi</div>),
        props: { onClickField: jest.fn() },
      },
    }];
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
 
  it('Should render without optionsActions - ok', () => {
    props.optionsActions = undefined;
    const component = shallow(getComponent(props));
    expect(component).toMatchSnapshot();
  });
  
  function getComponent(props) {
    return (<Item {...props} />);
  }
});
 
