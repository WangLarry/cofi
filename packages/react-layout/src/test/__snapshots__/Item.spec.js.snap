// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Should render - ok 1`] = `
<ItemView
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
  title="Employee"
/>
`;

exports[`Item Should render with all filtered sections - ok 1`] = `""`;

exports[`Item Should render with layout mobile - ok 1`] = `
<MobileItemView
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  size={1}
  title="Employee"
/>
`;

exports[`Item Should render with layout tabs - ok 1`] = `
<ItemView
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": undefined,
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
  title="Employee"
/>
`;

exports[`Item Should render with mainActions - popover - ok 1`] = `
<ItemView
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
  title="Employee"
/>
`;

exports[`Item Should render without layout - ok 1`] = `
<ItemView
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  title="Employee"
/>
`;

exports[`Item Should render without mainActions - ok 1`] = `
<ItemView
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
  title="Employee"
/>
`;

exports[`Item Should render without optionsActions - ok 1`] = `
<ItemView
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
  title="Employee"
/>
`;

exports[`Item Should render without title - ok 1`] = `
<ItemView
  sections={
    Array [
      Object {
        "id": "general",
        "ref": Object {
          "current": null,
        },
        "title": "General",
      },
    ]
  }
  sectionsRef={
    Object {
      "current": null,
    }
  }
  size={4}
  tabs={
    Object {
      "items": Array [
        Object {
          "label": "General",
          "value": "general",
        },
      ],
      "onChange": [Function],
      "value": "general",
    }
  }
/>
`;
